### ISSUE TITLE
test/integration/master package tests are slow, package times out

### ISSUE THREAD

**Issue by liggitt** [kubernetes/kubernetes#93609]
**Which jobs are flaking**:

pull-kubernetes-integration

**Which test(s) are flaking**:

test/integration/master

**Testgrid link**:

**Reason for failure**:

There are ~5 tests that average more than 30 seconds each.

https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-integration&graph-metrics=test-duration-minutes&include-filter-by-regex=k8s.io/kubernetes/test/integration/master&exclude-filter-by-regex=%5C%5B1%5C%5D

This is an umbrella issue tracking speeding up these five tests:

- [x] k8s.io/kubernetes/test/integration/master.TestReconcilerMasterLease{MultiMoreMasters,MultiCombined,Combined} - @gitlawr - https://github.com/kubernetes/kubernetes/pull/93721
- [x] k8s.io/kubernetes/test/integration/master.TestAudit - @vareti - https://github.com/kubernetes/kubernetes/pull/94054 
    * suggestion/starting point: rather than setting up a new server and webhook for each iteration, we can set up a single server with a policy that logs differently named configmaps at different verbosity levels (using the resourceNames portion of the audit policy), and a webhook that only intercepts specifically labeled configmaps (using the objectSelector portion of the webhook).
    * Then all the iterations in the table test can run against the same server and webhook, selecting a different configmap name depending on what verbosity we want to exercise, and setting different labels on the configmap depending on whether we want the mutating webhook to intercept it or not.
- [ ] k8s.io/kubernetes/test/integration/master.TestKMSHealthz - needs an owner

**Anything else we need to know**:

https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&text=FAIL%09k8s.io%2Fkubernetes%2Ftest%2Fintegration%2Fmaster

/sig api-machinery
/priority important-soon

**Comment by liggitt** [comment]
Possible approaches:
* when timeouts are configurable, speed them up (example in https://github.com/kubernetes/kubernetes/pull/93394)
* consider refactoring table tests to run in parallel (example in https://github.com/kubernetes/kubernetes/pull/93394)

**Comment by liggitt** [comment]
if you want to help improving the runtime of one of these tests, leave a comment, then link to a WIP PR once it is open

**Comment by liggitt** [comment]
cc @kubernetes/sig-api-machinery-bugs

**Comment by liggitt** [comment]
for TestAudit, rather than setting up a new server and webhook for each iteration, we can set up a single server with a policy that logs differently named configmaps at different verbosity levels (using the resourceNames portion of the audit policy), and a webhook that only intercepts specifically labeled configmaps (using the objectSelector portion of the webhook).

Then all the iterations in the table test can run against the same server and webhook, selecting a different configmap name depending on what verbosity we want to exercise, and setting different labels on the configmap depending on whether we want the mutating webhook to intercept it or not.

**Comment by vareti** [comment]
@liggitt I am fairly new to k8s and would like to get involved in this if I can. Any suggestions for which test to pick?

**Comment by liggitt** [comment]
@vareti sounds good, thanks! I added some initial thoughts in the description if you want some ideas. Feel free to reach out in slack in the #sig-apimachinery channel as you go

**Comment by gitlawr** [comment]
@liggitt I would like to work on master.TestReconcilerMasterLease. 
Seems like a lot of the time is used for starting apiservers, I think we can do that concurrently.

**Comment by liggitt** [comment]
@gitlawr, thanks I've put your name by that test

**Comment by apelisse** [comment]
cc @immutableT Are you interested in looking at the KMS test?

**Comment by ZhiFeng1993** [comment]
@liggitt I'd like to work on master.TestKMSHealthz

**Comment by liggitt** [comment]
thanks, I added your name. you might check with @immutableT to see if he has suggestions or has started looking at this

**Comment by ZhiFeng1993** [comment]
thanks @liggitt 

@immutableT please let me know if you've started looking. I'll let you have it. :)

**Comment by immutableT** [comment]
@ZhiFeng1993 I have not yet.
If you are interested in working on this then by all means go ahead.

**Comment by ZhiFeng1993** [comment]
@immutableT  I looked into `TestKMSHealthz` and I haven't found a good way to speed up the test according to the strategies @liggitt suggested: the activities are in sequence so I couldn't really parallelize anything. Also I have't found any timeout defined. 

Any thoughts/suggestions?

**Comment by vareti** [comment]
> a policy that logs differently named configmaps at different verbosity levels (using the resourceNames portion of the audit policy)

@liggitt `resourceNames` does not seem to be working with `CREATE` verb for admission webhooks. I am thinking of using different resource type for different verbosity level. Does that sound good? Please let me know if there is a simple alternative.

**Comment by liggitt** [comment]
> @liggitt `resourceNames` does not seem to be working with `CREATE` verb for admission webhooks. I am thinking of using different resource type for different verbosity level. Does that sound good? Please let me know if there is a simple alternative.

isolating each test to its own namespace (instead of using `default` for all of them) and filtering the audit rules using the `namespaces` portion of the rule would work

**Comment by fedebongio** [comment]
/assign @jiahuif 
Thought could be a good learning experience, I'm sure you'll pull this one out
/triage accepted

**Comment by BenTheElder** [comment]
The only outstanding test is sig-auth (`k8s.io/kubernetes/test/integration/master.TestKMSHealthz`) so:
/remove-sig api-machinery
/sig-auth
/triage accepted

also #126202

**Comment by neolit123** [comment]
/sig api-machinery

i though the master->controlplane wg-naming work cleaned these parts. 🤔

**Comment by BenTheElder** [comment]
We believe this remaining test case is owned by SIG auth
/sig auth
/remove-sig api-machinery
WG naming didn't see through a lot of renaming.

**Comment by BenTheElder** [comment]
https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-integration&graph-metrics=test-duration-minutes&include-filter-by-regex=k8s.io/kubernetes/test/integration/controlplane&exclude-filter-by-regex=%5C%5B1%5C%5D

https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-integration&graph-metrics=test-duration-minutes

this is still one of the slower tests, but it's not timing out.
x-ref other changes in https://github.com/kubernetes/kubernetes/issues/126202

**Comment by BenTheElder** [comment]
probably best to track in a new issue at this point.
